### ISSUE TITLE
vacancy/solute sites in HCP Ru

### ISSUE THREAD

**Issue by DS-Wen** [pydii/pydii#4]
Hi developers,

I got an error message when I try to create vacancy/solute sites in HCP Ru (MaterialsProject id mpid-33).

When I used
```gen_sol_pref_structure --mpid mp-33 --mapi_key xxxxx --solute Co```

It showed the message: 
```
Traceback (most recent call last):  
  File "/opt/anaconda3/bin/gen_sol_pref_structure", line 11, in <module>  
    load_entry_point('pydii==1.1.0', 'console_scripts', 'gen_sol_pref_structure')()  
  File "/opt/anaconda3/lib/python3.7/site-packages/pydii-1.1.0-py3.7.egg/pydii/scripts/gen_def_structure.py", line 275, in im_sol_sub_def_struct_gen
    substitute_def_struct_gen(args.mpid, args.solute, args.mapi_key, args.cellmax)
  File "/opt/anaconda3/lib/python3.7/site-packages/pydii-1.1.0-py3.7.egg/pydii/scripts/gen_def_structure.py", line 179, in substitute_def_struct_gen
    vac = Vacancy(structure=prim_struct, defect_site=site)
  File "/opt/anaconda3/lib/python3.7/site-packages/pymatgen/analysis/defects/core.py", line 59, in __init__
    self._multiplicity = multiplicity if multiplicity else self.get_multiplicity()
  File "/opt/anaconda3/lib/python3.7/site-packages/pymatgen/analysis/defects/core.py", line 188, in get_multiplicity
    raise ValueError("Site {} is not in bulk structure! Cannot create Vacancy object.".format(self.site))
ValueError: Site [ 0.         0.        -2.1569615] Ru is not in bulk structure! Cannot create Vacancy object. 
```

I soon realized that in the
 ```substitute_def_struct_gen(mpid, solute, mapi_key, cellmax)```
the way to enumerate the symmetrically distinct vacancy sites can cause this error. 
The ```prim_struct2```  somehow has an atomic site not in the HCP primitive cell.

Thanks,
Dongsheng

**Comment by mbkumar** [comment]
The underlying defect classes I wrote in pymatgen changed quite a bit. I'll
investigate this and get back to you.

Bharat Medasani

Engineer
Princeton Plasma Physics Lab (PPPL)

**Comment by enze-chen** [comment]
Hi Dongsheng, thanks for pointing this out. I'll try to take a look later this week as well. I had to update this section because, as Bharat points out, Pymatgen has changed a lot in the past six years. Apparently my fix was not robust enough. :/

**Comment by enze-chen** [comment]
Hello @DS-Wen ,
Sorry about the late reply. I took a look and implemented a fix in #6 . I actually overcomplicated the original code and the new version is more precise. I tested it with a combination of mpid's, user inputs, and different crystal structures (including your HCP example). The results seem sensible to me, but you should verify. 
Best,
Enze
